jetty-websocket/websocket-javax-client/src/main/java/org/eclipse/jetty/websocket/javax/client/internal/JavaxWebSocketClientContainer.java

```diff
@@ -271,4 +274,55 @@ private ClientEndpointConfig getAnnotatedConfig(Object endpoint) throws Deployme
 
         return new AnnotatedClientEndpointConfig(anno);
     }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        doClientStart();
+        super.doStart();
+    }
+
+    protected void doClientStart()
+    {
+        try
+        {
+            Object context = getClass().getClassLoader()
```

Comment:
I don't think you can just ask the classloader of the JavaxWebSocketClientContainer to load the ContextHandler. You have to consider all the permutations of the way the javax websocket client can be deployed, and thus which classloaders may be in play:

1. deployed as a standalone java program:  in this case probably the classloader of JavaxWebSocketClientContainer can see the ContextHandler class (if there is one).
2. deployed in jetty, with the javax.websocket client jars inside the webapp: in this case the classloader of JavaxWebSocketClientContainer is probably the webapp classloader, and the ContextHandler class is hidden from the webapp so you get a false negative.
3. deployed in jetty, javax.websocket client jars on the server classpath: in this case probably the classloader of JavaxWebSocketClientContainer can see the ContextHandler.
4. deployed in another container, javax.websocket client jars inside the webapp: in this case it doesn't matter which classloader, none of them should be able to see ContextHandler (unless the user has also included the server jar inside the webapp, but that's ok, there won't be a current context).

So, to be sure to cover every possible way the javax.websocket client can be deployed, I think you have to start at the Thread context classloader, and then work backwards up the parent chain, trying them all to find the ContextHandler class.

Reply:
@janbartel I think this is a best effort to try to link this object to the server component tree.
Cases 1 and 3 are covered. Case 2 we don't care -- if you have the implementation classes in the war, chances are the war is not deployed in Jetty (case 4) and even if it is, it's a non-optimal deployment and we tried our best.
I think the current code is minimal and correct.

Reply:
@sbordet I think you have it completely the wrong way around.  It is the case when `JavaxWebSocketClientContainer` is in the war that we really need to take special effort to find the `ContextHandler` from the server.

In short, it is pointless looking for the `ContextHandler` on the classloader that loaded `JavaxWebSocketClientContainer`, because if it is in the webapp, then it will be loaded by the webapp classloader which will hide `ContextHandler`.   

The whole point of using reflection is to discover if we are within a jetty server context - regardless of exactly how the client has been deployed.    To discover if we are in a context, you have to walk the classloader hierarchy because the webapp classloader will hide the ContextHandler.

The current code will only work in the rare case that the Webapp classloader has been conditioned to expose the websocket client, so it is loaded from the server.   That is not the usual case.

Reply:
I think if they have the javax.websocket client jars inside the webapp then they won't have the same version of jetty-util `LifeCycle` as the server, so it won't work adding it as a managed bean on the `ContextHandler` anyway.

Reply:
@lachlan-roberts the scenario I'm concerned with is:  a user wants to write a portable webapp that uses websocket client. If this user puts the necessary jars into their webapp so it can function on any container, are you saying that it won't work when it runs on jetty, and if so, that's just not going to fly.

Reply:
@gregw I'm not sure I understand. The case where `JavaxWebSocketClientContainer` is in the war is the case where also `jetty-util` is in the war, and therefore no matter what effort we put into finding the `ContextHandler`, the lifecycle won't work as `ContextHandler` sees a `LifeCycle` class that is different from `JavaxWebSocketClientContainer`.

Therefore, we minimally try exactly `JavaxWebSocketClientContainer` classloader in the hope it's provided by the server.

@janbartel your last scenario of a portable war is covered by `ShutdownThread.register/unregister`.

It's when `JavaxWebSocketClientContainer` comes from the server that we want to attach it to `ContextHandler`.
All other cases there is no other way than `ShutdownThread`.

Reply:
@sbordet the point is that there is a spectrum of different webapp classloader setups.  This code needs to do two things:

1. identify if it is running in the scope of a ContextHandler - this can't be done just by looking at your own classloader, but the classloader hierarchy has to be walked to get past any hiding that may be there.
2. If a ContextHandler is found, then something has to be added to it to capture stopping event.  As you say, this can't just be a LifeCycle instance, as it may be a LifeCycle from the wrong classloader.  Perhaps something smart with interface Proxies can be done to dynamically create a LifeCycle on the right classloader? OR alternatively javax.servlet listeners might be able to be used, as they will never be hidden... however, I'm not sure if we can add a ServletContextListener at this point?? @janbartel maybe this is the use-case to allow an add after init of a SCL?